services/drupal/hash.go

```diff
@@ -0,0 +1,58 @@
+package drupal
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"strings"
+)
+
+// generateRandomBytes returns securely generated random bytes.
+// It will return an error if the system's secure random
+// number generator fails to function correctly, in which
+// case the caller should not continue.
+func generateRandomBytes(n int) ([]byte, error) {
+	b := make([]byte, n)
+	_, err := rand.Read(b)
+	// Note that err == nil only if we read len(b) bytes.
+	if err != nil {
+		return nil, err
+	}
+
+	return b, nil
+}
+
+// GenerateRandomHashSalt returns a securely generated random string.
+// It will return an error if the system's secure random
+// number generator fails to function correctly, in which
+// case the caller should not continue.
+func GenerateRandomHashSalt() (string, error) {
+	assertAvailablePRNG()
+	const letters = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-"
+	bytes, err := generateRandomBytes(55)
+	if err != nil {
+		return "", err
+	}
+	for i, b := range bytes {
+		bytes[i] = letters[b%byte(len(letters))]
+	}
+
+	s := base64.URLEncoding.EncodeToString(bytes)
+	s = strings.Replace(s, "+", "-", -1)
+	s = strings.Replace(s, "/", "-", -1)
+	s = strings.Replace(s, "=", "", -1)
+
+	return s, err
```

Comment:
Not sure what this salt generation function is for exactly. But if there's a risk that the process will fail due to not having random number generation capability, maybe should try a different tack? I'm not sure how common this case would be though?

For a random string, one that would look like an API key from most vendors, could generate a HMAC SHA1 from a secret of the current datetime, then converted to base64

Reply:
I think it should never happen that the user can't generate random crypto. In such a case, things like lastpass wouldn't even work. But, _if_ it happens, I think we should error, since none of the cryptographic features of the OS would work properly and we'd be giving the false impression of having a secure hash when we don't. 

The hash, FYI, is used by drupal and has funny and weird rules around it's construction. Which is why we specifically replace things like `=` and `/`.